File: Subsets.py
class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
    def subsets(self, S):
        res = [[]]
        S.sort()
        for e in S:
            res = res+[l+[e] for l in res]
        return res

def subsets(S):
    A = [[]]
    S.sort()
    for n in S:
        for i in range(len(A)):
            ss = A[i][:]
            ss.append(n)
            A.append(ss)
    return A


if __name__ == "__main__":
    print Solution().subsets([1,2,3])
    print subsets([1,2,3])

File: Subsets.cpp
/*
 * subsets.cpp
 * Copyright (C) 2014 fuel <fuel@fuel-3330>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    vector<vector<int> > subsets(vector<int> &S) {
        vector<vector<int> > res(1);
        sort(S.begin(),S.end());
        for(auto e : S){
            int n = res.size();
            for (int i = 0; i < n; ++i) {
                vector<int> l(res[i]);
                l.push_back(e);
                res.push_back(l);
            }
        }
        return res;
    }
};
